internal/gatewayapi/validate.go

```diff
@@ -240,111 +241,126 @@ func (t *Translator) validateTLSConfiguration(listener *ListenerContext, resourc
-			) {
+			secret := resources.GetSecret(secretNamespace, string(certificateRef.Name))
+
+			if secret == nil {
 				listener.SetCondition(
 					v1beta1.ListenerConditionResolvedRefs,
 					metav1.ConditionFalse,
-					v1beta1.ListenerReasonRefNotPermitted,
-					fmt.Sprintf("Certificate ref to secret %s/%s not permitted by any ReferenceGrant", *certificateRef.Namespace, certificateRef.Name),
+					v1beta1.ListenerReasonInvalidCertificateRef,
+					fmt.Sprintf("Secret %s/%s does not exist.", listener.gateway.Namespace, certificateRef.Name),
 				)
 				break
 			}
 
-			secretNamespace = string(*certificateRef.Namespace)
-		}
-
-		secret := resources.GetSecret(secretNamespace, string(certificateRef.Name))
+			if secret.Type != v1.SecretTypeTLS {
+				listener.SetCondition(
+					v1beta1.ListenerConditionResolvedRefs,
+					metav1.ConditionFalse,
+					v1beta1.ListenerReasonInvalidCertificateRef,
+					fmt.Sprintf("Secret %s/%s must be of type %s.", listener.gateway.Namespace, certificateRef.Name, v1.SecretTypeTLS),
+				)
+				break
+			}
 
-		if secret == nil {
-			listener.SetCondition(
-				v1beta1.ListenerConditionResolvedRefs,
-				metav1.ConditionFalse,
-				v1beta1.ListenerReasonInvalidCertificateRef,
-				fmt.Sprintf("Secret %s/%s does not exist.", listener.gateway.Namespace, certificateRef.Name),
-			)
-			break
-		}
+			if len(secret.Data[v1.TLSCertKey]) == 0 || len(secret.Data[v1.TLSPrivateKeyKey]) == 0 {
+				listener.SetCondition(
+					v1beta1.ListenerConditionResolvedRefs,
+					metav1.ConditionFalse,
+					v1beta1.ListenerReasonInvalidCertificateRef,
+					fmt.Sprintf("Secret %s/%s must contain %s and %s.", listener.gateway.Namespace, certificateRef.Name, v1.TLSCertKey, v1.TLSPrivateKeyKey),
+				)
+				break
+			}
 
-		if secret.Type != v1.SecretTypeTLS {
-			listener.SetCondition(
-				v1beta1.ListenerConditionResolvedRefs,
-				metav1.ConditionFalse,
-				v1beta1.ListenerReasonInvalidCertificateRef,
-				fmt.Sprintf("Secret %s/%s must be of type %s.", listener.gateway.Namespace, certificateRef.Name, v1.SecretTypeTLS),
-			)
-			break
-		}
+			pka, err := validateTLSSecretData(secret)
+			if err != nil {
+				listener.SetCondition(
+					v1beta1.ListenerConditionResolvedRefs,
+					metav1.ConditionFalse,
+					v1beta1.ListenerReasonInvalidCertificateRef,
+					fmt.Sprintf("Secret %s/%s must contain valid %s and %s, %s.", listener.gateway.Namespace, certificateRef.Name, v1.TLSCertKey, v1.TLSPrivateKeyKey, err.Error()),
+				)
+				break
+			}
 
-		if len(secret.Data[v1.TLSCertKey]) == 0 || len(secret.Data[v1.TLSPrivateKeyKey]) == 0 {
-			listener.SetCondition(
-				v1beta1.ListenerConditionResolvedRefs,
-				metav1.ConditionFalse,
-				v1beta1.ListenerReasonInvalidCertificateRef,
-				fmt.Sprintf("Secret %s/%s must contain %s and %s.", listener.gateway.Namespace, certificateRef.Name, v1.TLSCertKey, v1.TLSPrivateKeyKey),
-			)
-			break
+			// Check whether the public key algorithm in the referenced secrets are unique.
```

Comment:
do we need to enforce this ?
what about the case when `algo` is same but the `SNI` is different

Reply:
I haven't covered multiple certs and SNI based matching in this PR. Maybe we can keep that as separate? 
Earlier we had a hard enforcement of 1 cert per listener, which this PR relaxes, and allows for 2 now - with varying algos.

Reply:
IIUC right now, people can configure different hosts in different listeners, with specific tls termination configs? and then with this PR they'll be able to configure dual certs for those hosts

Reply:
cool, as suggested can you raise a GH issue to relax this for multi cert SNI selection ?

Reply:
seems like envoy should support it based on https://github.com/envoyproxy/envoy/issues/24824